agent/app/agent.go

```diff
@@ -231,6 +234,13 @@ func newAgent(blackholeEC2Metadata bool, acceptInsecureCert *bool) (agent, error
 		metadataManager = containermetadata.NewManager(dockerClient, cfg)
 	}
 
+	credentialProviderV2 := awsv2.NewCredentialsCache(
```

Comment:
Do we see a scenario where `InstanceCredentialsProvider` can be needed without cache? Can `NewCredentialsCache` be pushed inside `NewInstanceCredentialsProvider`, so that credentials provider is always cached?

Reply:
Yes, when we construct our SDK v2 clients, they will pass/use a credentials provider to/from `LoadDefaultConfig`. From the [docs](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specify-credentials-programmatically),
> All credential providers passed to or returned by LoadDefaultConfig are wrapped in a [CredentialsCache](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#CredentialsCache) automatically.

The code would look something like this:
```
cfg, err := config.LoadDefaultConfig(
    context.TODO(),
    config.WithCredentialsProvider(provider),
)
if err != nil {
    return err
}
ecsClient := ecs.NewFromConfig(cfg)
```

Reply:
https://github.com/aws/amazon-ecs-agent/pull/4424/commits/a9b6e6de25541353ec5c7c6d640ed21374a08ca7 updates the constructor to wrap the provider in a cache and renames `InstanceCredentialsProvider` to `InstanceCredentialsCache`.